User Management Page

jabbate19 opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Adds a User Management page. Instead of needing to use the GraphQL Playground, you can approve/promote users in the website.

#### Which issue(s) this PR fixes:
N/A

jabbate19 commented:
![image](https://github.com/spellshift/realm/assets/40615740/d4b0cdaa-b000-46f4-8ebe-5b04bf1aa4a4)

cmp5987 commented on tavern/internal/www/src/components/page-wrapper/PageWrapper.tsx:
```diff
@@ -21,6 +22,7 @@ const navigation = [
   { name: PageNavItem.createQuest, href: '/createQuest', icon: CommandLineIcon, internal: true },
   { name: PageNavItem.dashboard, href: '/dashboard', icon: PresentationChartBarIcon, internal: true },
   { name: PageNavItem.hosts, href: '/hosts', icon: BugAntIcon, internal: true },
+  { name: PageNavItem.users, href: '/users', icon: UserIcon, internal: true },
```
We only want to display this view if a user is an Admin.

Also we should change the endpoint to something like /admin-settings

cmp5987 commented on tavern/internal/www/src/pages/user-list/components/UserTable.tsx:
```diff
@@ -0,0 +1,92 @@
+import { Badge, Image, Button } from "@chakra-ui/react";
```
Use tavern-ui version of badge and button instead of chakra-ui

cmp5987 commented on tavern/internal/www/src/pages/user-list/components/UserTable.tsx:
```diff
@@ -0,0 +1,92 @@
+                            </div>
+                        </Badge>}
+                        {!rowData?.isActivated && <Badge>
+                            <div className="flex flex-row gap-1 justify-center items-center p-1" >
```
This nested div doesn't seem needed

cmp5987 commented on tavern/internal/www/src/pages/user-list/components/UserTable.tsx:
```diff
@@ -0,0 +1,92 @@
+                const rowData = cellData.getValue();
+                return (
+                    <div className="flex flex-row flex-wrap gap-1">
+                      {!rowData?.isActivated && <Button size={"sm"}>
```
Looks like you are missing onClick for something to occur with these buttons?

cmp5987 commented on tavern/internal/www/src/pages/user-list/hooks/useUserTable.ts:
```diff
@@ -0,0 +1,38 @@
+import { convertArrayOfObjectsToObject, getOfflineOnlineStatus } from "../../../utils/utils";
+
+export const useUserTable = () => {
+    const [users, setUsers] = useState([]);
```
Not sure you need this.

Instead you could just return data?.users || []

cmp5987 commented on tavern/internal/www/src/utils/enums.ts:
```diff
@@ -16,6 +16,7 @@ export enum PageNavItem {
     tasks='Tasks',
     createQuest= 'Create new quest',
     hosts="Hosts",
+    users="User Management",
```
We should change these to be related to admin settings since we expect more kinds of admin controls on this page in the future.

cmp5987 commented on tavern/internal/www/src/pages/user-list/components/UserTable.tsx:
```diff
@@ -0,0 +1,92 @@
+    ];
+
+    return (
+        <div>
```
Not sure you need this extra div?

cmp5987 reviewed (commented):
Just a few comments based on what you currently have written.

jabbate19 commented:
Updated photos

Admin UI

<img width="1691" alt="image" src="https://github.com/user-attachments/assets/5ad5f425-97c8-4552-9514-4864a07ac7e5" />

If a user attempts to access /admin while not admin

<img width="630" alt="image" src="https://github.com/user-attachments/assets/a8f014c5-5b65-4804-abd3-bb7aba164298" />

Non-Admin Users still do not see Admin in navbar

<img width="326" alt="image" src="https://github.com/user-attachments/assets/5d3f975f-1882-45ae-b919-4c59287e6ecb" />

cmp5987 commented on tavern/internal/www/src/components/page-wrapper/PageWrapper.tsx:
```diff
@@ -34,13 +38,35 @@ function classNames(...classes: string[]) {
+
+  const shownNavigation = navigation.filter((item) => data!.me!.isAdmin || !item.adminOnly);
+
+  console.log(shownNavigation)
```
Looks like a left over console.log

cmp5987 commented on tavern/internal/www/src/components/page-wrapper/PageWrapper.tsx:
```diff
@@ -34,13 +38,35 @@ function classNames(...classes: string[]) {
   const [sidebarOpen, setSidebarOpen] = useState(false)
+  const {data, isLoading, error} = useContext(AuthorizationContext);
+
+  if(isLoading){
```
I think it would be good to pull this logic outside of PageWrapper so the PageWrapper is only accessed after AuthorizationContext is loaded.

jabbate19 replied:
I kept this here because it needs to be used by AccessGate and Navigation. Is there an easy way to use it in both places without using context twice?

cmp5987 commented on tavern/internal/www/src/pages/admin/hooks/userMutations.ts:
```diff
@@ -0,0 +1,54 @@
+        }
+    `;
+
+    const handleError = (error: NetworkError | GraphQLErrors) => {
```
When are these errors displayed, I saw the submit used by the table. 

Additionally, How are they reset to false for if a user tries to re-trigger the action on a failure?

jabbate19 replied:
Wdym by this?

cmp5987 replied:
When an error occurs as a result of a user demoting/promoting a user. We should be displaying these errors to the user. This could occur via Toast or other banner message. It should also be dismissible.

cmp5987 commented on tavern/internal/www/src/pages/admin/AdminPortal.tsx:
```diff
@@ -0,0 +1,27 @@
+                    <span>This portal is for managing users allowed to access Tavern. You can also promote users to Administrator, who will be able to access this page.</span>
+                </div>
+            </div>
+            <Tab.Group>
```
Any plans for what this will expand to include, right now we have only 1 tab so its a bit odd to include a tab/tab panel.

jabbate19 replied:
Since we are naming it an "Admin" panel, thought it would be good to add tabs if other admin features need to get added later

cmp5987 commented on tavern/internal/www/src/pages/admin/components/UserTable.tsx:
```diff
@@ -0,0 +1,91 @@
+                      {!rowData?.isActivated && <Button disabled={isDisabled} buttonStyle={{ color: "gray", size: "md" }} onClick={() => submitUpdateUser({"id": rowData.id, "activated": true, "admin": false})}>
+                        Activate
+                      </Button>}
+                      {rowData?.isActivated && <Button disabled={isDisabled} buttonStyle={{ color: "gray", size: "md" }} onClick={() => submitUpdateUser({"id": rowData.id, "activated": false, "admin": false})}>
```
We should make 'Deactivate' a 'red' color. Traditionally, any 'destructive' action is usually 'red'. Same can be done with Demote.

cmp5987 commented on tavern/internal/www/src/pages/admin/AdminPortal.tsx:
```diff
@@ -0,0 +1,27 @@
+    return (
+        <PageWrapper currNavItem={PageNavItem.admin}>
+            <div className="border-b border-gray-200 pb-5 sm:flex sm:items-center sm:justify-between">
+                <h3 className="text-xl font-semibold leading-6 text-gray-900">Admin Portal</h3>
```
The styling of this is not matching other headers. 
Can you match the styling of other pages where the details/paragraph explaining the page are flex flex-col.

<img width="688" alt="Screenshot 2025-01-24 at 8 47 41 PM" src="https://github.com/user-attachments/assets/b4c2e32f-3635-40db-93f7-903d8e3bf0f4" />

jabbate19 commented:
<img width="987" alt="image" src="https://github.com/user-attachments/assets/cd6aa46e-a76a-42ce-8cd6-1722c8fe2f57" />
<img width="634" alt="image" src="https://github.com/user-attachments/assets/f62f3925-1bca-4ab2-9655-284be89766f7" />

New red button styles

<img width="707" alt="image" src="https://github.com/user-attachments/assets/bb58076b-5fd9-4782-a82d-d09607eb54b8" />
Fixed text location

jabbate19 commented:
<img width="992" alt="image" src="https://github.com/user-attachments/assets/41a13916-de1d-440d-8766-13d6d630abf3" />

Without tabs

jabbate19 commented:
Solves #838

KCarretto commented:
Will leave this review to @cmp5987  as well

cmp5987 commented:
Nits: 
- Mind adding 'buttonVariant="ghost" to the action buttons. 
- Add feedback handling for success or error cases when users click 'actions. Think Toast or Inline banners that are dismissible. 
- Don't forget to squash before merging but this looks almost ready.
- Add loading states to buttons and prevent interaction when a user's action is being processed. In scenarios where backend is slow we don't really want to have things double clicked or users to not understand that their action is getting processed. 

Task Output vs Task Finished
I think the task output vs task finished really depends. The ideal for the statistic would be tasks with meaningful information available. Unfortunately, 'finished' doesn't mean success/not failed. Output can also be filled with errors depending on the task V_V. Let's stick with whichever you prefer and iterate based on 2025 ISTS feedback.

jabbate19 commented:
> Add feedback handling for success or error cases when users click 'actions. Think Toast or Inline banners that are dismissible.

Is there an example in the codebase I can use?

jabbate19 commented:
> Mind adding 'buttonVariant="ghost" to the action buttons.

<img width="1044" alt="image" src="https://github.com/user-attachments/assets/23d853b7-19d6-43fe-afcf-f971fe60b937" />

Your call, but imo I think this looks off unless we add a background color to the button or outline

cmp5987 commented:
> > Add feedback handling for success or error cases when users click 'actions. Think Toast or Inline banners that are dismissible.
> 
> Is there an example in the codebase I can use?

For an Inline banner example we have 'AlertError' component that was used in our Tome repo setup.
https://github.com/spellshift/realm/blob/main/tavern/internal/www/src/pages/tomes/components/StepCreateRepository.tsx#L32

This is a simple guide on using Chakra-UI Toasts for user feedback.
https://medium.com/@santhoshs19032003/easy-toast-messages-in-react-with-chakra-ui-a-beginners-guide-8833807d9e12

jabbate19 commented:
- Changed button to outline, imo looks better with the outline contract than white on white
- Added toasts

<img width="827" alt="image" src="https://github.com/user-attachments/assets/c627221e-a931-44af-9d68-ea837e0245dd" />
<img width="1041" alt="image" src="https://github.com/user-attachments/assets/75d21d1c-32b1-441a-ab14-58a459fd78bd" />
<img width="1020" alt="image" src="https://github.com/user-attachments/assets/b06ec8ee-db7e-495c-827e-62366b5985f1" />
<img width="1013" alt="image" src="https://github.com/user-attachments/assets/9d66adf7-b90f-4962-b9c2-59f626391f65" />

cmp5987 reviewed (approved):
Awesome work, 
Don't forget to squash, rebase and rebuild the UI.

~ I think we can squeeze this in for next weeks release.
